atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionScopedCacheImplTest.java

```diff
@@ -138,6 +152,186 @@ public void allValuesReadFromCachePreventsReadToDb() {
+                        CacheEntry.unlocked(CacheValue.empty()),
+                        CellReference.of(TABLE, createCell(1, 2)),
+                        CacheEntry.unlocked(CacheValue.empty()),
+                        CellReference.of(TABLE, createCell(1, 3)),
+                        CacheEntry.unlocked(CacheValue.empty()),
+                        CellReference.of(TABLE, createCell(2, 1)),
+                        CacheEntry.unlocked(CacheValue.empty()),
+                        CellReference.of(TABLE, createCell(2, 3)),
+                        CacheEntry.unlocked(CacheValue.empty())),
+                HashSet.of(TABLE),
+                ImmutableSet.of(TABLE));
+        TransactionScopedCache cache = TransactionScopedCacheImpl.create(snapshot, metrics);
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(1), createBytes(2), createBytes(3));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        Multiset<Cell> directLookups = HashMultiset.create();
+        NavigableSet<byte[]> rowLookups = new TreeSet<>(UnsignedBytes.lexicographicalComparator());
+
+        NavigableMap<byte[], RowResult<byte[]>> lookup =
+                cache.getRows(TABLE, rowsAndCols, columns, cells -> loadCells(cells, directLookups), rows -> {
+                    rows.forEach(rowLookups::add);
+                    return new TreeMap<>(UnsignedBytes.lexicographicalComparator());
+                });
+
+        Cell onlyCell = createCell(2, 2);
+        assertThat(directLookups).containsExactly(onlyCell);
+        assertThat(rowLookups).containsExactly(createBytes(3));
+        assertThat(lookup.size()).isEqualTo(1);
+        assertThat(lookup.get(createBytes(2)).getOnlyColumnValue()).containsExactly(createBytes(onlyCell));
+    }
+
+    @Test
+    public void getRowsDoesNotBreakOnEmptyRowsAndCachesThemCorrectly() {
+        TransactionScopedCache cache = TransactionScopedCacheImpl.create(snapshotWithSingleValue(), metrics);
+
+        ImmutableList<byte[]> rowsAndCols = ImmutableList.of(createBytes(2), createBytes(3));
+        ColumnSelection columns = ColumnSelection.create(rowsAndCols);
+
+        Multiset<Cell> directLookups = HashMultiset.create();
+
+        AtomicInteger rowLookups = new AtomicInteger();
+        Function<Iterable<byte[]>, NavigableMap<byte[], RowResult<byte[]>>> rowLookup = rows -> {
+            rowLookups.incrementAndGet();
+            return new TreeMap<>(UnsignedBytes.lexicographicalComparator());
+        };
+        NavigableMap<byte[], RowResult<byte[]>> lookup =
+                cache.getRows(TABLE, rowsAndCols, columns, cells -> loadCells(cells, directLookups), rowLookup);
+        cache.getRows(TABLE, rowsAndCols, columns, cells -> loadCells(cells, directLookups), rowLookup);
+
+        assertThat(directLookups).isEmpty();
+        assertThat(lookup).isEmpty();
+        assertThat(rowLookups).hasValue(1);
+    }
```

Comment:
why is rowLookups 1 and not 2?

Reply:
This was a bit misleading. It gets incremented once for each getRows call. The first time we call it once, for two rows, the second time it is not called since everything has been cached. Refactored to be more clear